xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/doc/XWikiDocument.java

```diff
@@ -4093,6 +4094,20 @@ public void readFromForm(EditForm eform, XWikiContext context) throws XWikiExcep
         readDocMetaFromForm(eform, context);
         readTranslationMetaFromForm(eform, context);
 
+        readAddedUpdatedAndRemovedObjectsFromForm(eform, context);
+    }
+
+    /**
+     * Adds objects, applies property updates and removes objects as specified in the form.
```

Comment:
So I noticed that we don't have a javadoc for `#readObjectsFromForm`: so right now I have a hard time to understand what's the difference between this new method and `readObjectsFromForm`, it could be good to explain this difference at least in this javadoc, to understand when to use one and when to use the others, since it's not clear right now.

Reply:
I agree that this is hard to understand. I didn't want to fundamentally change the behavior of `readObjectsFromForm` by including this additional functionality, in particular as the new method calls either `readObjectsFromForm` or `readObjectsFromFormUpdateOrCreate` depending on the `objectPolicy` request parameter. In my opinion, already `readObjectsFromFormUpdateOrCreate` should have been an extension of `readObjectsFromForm`. If you think the code organization should be changed, please let me know.

I cannot tell you when to use one or the other, in my opinion it should always be `readAddedUpdatedAndRemovedObjectsFromForm` as it is the one method that handles all object updates. I could imagine that there might be scenarios where this is not desirable, e.g., if you rely on the fact that objects you specified for deletion are still there during a preview such that you can display the properties of the objects that would be deleted. However, these examples seem so far-fetched and would require such a lot of custom code, e.g., for re-displaying the form for deleting the objects for example that I find it hard to imagine that somebody relies on this behavior.

Reply:
> I cannot tell you when to use one or the other

Hmmm I've the feeling that `readAddedUpdatedAndRemovedObjectsFromForm` code was created for some specific requests, e.g. for partial document requests when you only provide a part of the information you want to update, and not all of them. It probably evolved to support more usecases. While on the other side, `readObjectsFromForm` was done to work with the all objects infromation being sent. 
Maybe we should start deprecating `readObjectsFromForm` in favor of the other one, if it supports as well all scenario.

Reply:
> Hmmm I've the feeling that `readAddedUpdatedAndRemovedObjectsFromForm` code was created for some specific requests, e.g. for partial document requests when you only provide a part of the information you want to update, and not all of them.

According to the git history, you added that code for adding and removing objects as part of [XWIKI-16210](https://jira.xwiki.org/browse/XWIKI-16210). The code in `readObjectsFromFormUpdateOrCreate` is from [XWIKI-10252](https://jira.xwiki.org/browse/XWIKI-10252). All methods support updating just some objects and do not remove anything if a property is not specified in the request, though `readObjectsFromFormUpdateOrCreate` also cleans-up properties that do not belong to the xclass. In my opinion, we should always process all object information from the form (i.e., also added and removed objects) to avoid issues like the one I'm fixing here but it seems also strange to now merge everything into the one method `readObjectsFromForm` which is the original method for updating objects from the form.